### ISSUE TITLE
passing parameters to jax.jit 'ed functions

### ISSUE THREAD

**Issue by mgbukov** [google/jax#1922]
The following script throws an error (see below for traceback) if the `jit` decorator is on. When I comment the jit-decorator out, it runs properly. 

If this is deliberate, could someone clarify what causes this behavior?  

```
from jax import jit
import jax.numpy as jnp
import numpy as np

@jit
def test(inputs,axis=(0,)):
    mean=jnp.mean(inputs,axis=axis,keepdims=True)

input_shape=(8,16)
inputs=np.random.uniform(size=input_shape)
test(inputs,axis=(0,))
```

Here's the traceback with the jit decorator on:

```
Traceback (most recent call last):
  File "bug.py", line 13, in <module>
    test(inputs,axis=(0,))
  File "/Users/mgbukov/jax/jax/api.py", line 148, in f_jitted
    out = xla.xla_call(flat_fun, *args_flat, device=device, backend=backend)
  File "/Users/mgbukov/jax/jax/core.py", line 569, in call_bind
    outs = primitive.impl(f, *args, **params)
  File "/Users/mgbukov/jax/jax/interpreters/xla.py", line 378, in _xla_call_impl
    compiled_fun = _xla_callable(fun, device, backend, *map(abstractify, args))
  File "/Users/mgbukov/jax/jax/linear_util.py", line 217, in cached_fun
    ans, f_prev = cached_fun_body(f, args)
  File "/Users/mgbukov/jax/jax/linear_util.py", line 214, in cached_fun_body
    return call(f, *args), f
  File "/Users/mgbukov/jax/jax/interpreters/xla.py", line 394, in _xla_callable
    jaxpr, (pvals, consts, env) = pe.trace_to_subjaxpr(fun, master, False).call_wrapped(pvals)
  File "/Users/mgbukov/jax/jax/linear_util.py", line 165, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "bug.py", line 7, in test
    mean=jnp.mean(inputs,axis=axis,keepdims=True)
  File "/Users/mgbukov/jax/jax/numpy/lax_numpy.py", line 1100, in mean
    normalizer = onp.prod(onp.take(shape(a), axis))
  File "<__array_function__ internals>", line 6, in take
  File "/Users/mgbukov/miniconda3/envs/jax/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 194, in take
    return _wrapfunc(a, 'take', indices, axis=axis, out=out, mode=mode)
  File "/Users/mgbukov/miniconda3/envs/jax/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 58, in _wrapfunc
    return _wrapit(obj, method, *args, **kwds)
  File "/Users/mgbukov/miniconda3/envs/jax/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 47, in _wrapit
    result = getattr(asarray(obj), method)(*args, **kwds)
ValueError: setting an array element with a sequence.
```

**Comment by clemisch** [comment]
You have to declare the `axis` parameter as static via the `static_argnums` argument of `jit`:

```python
from jax import jit
import jax.numpy as jnp
import numpy as np
import jax

@jax.partial(jit, static_argnums=1)  # jit with axis being static
def test(inputs, axis):  # no default value for axis as `test` needs to be called with all arguments
    mean=jnp.mean(inputs,axis=axis,keepdims=True)

input_shape=(8,16)
inputs=np.random.uniform(size=input_shape)
test(inputs, (0,))  # `test` needs to be called with all arguments
```

I marked the changes with comments.

**Comment by mgbukov** [comment]
this was very useful, thanks!

A few follow-up Q's:

* is there a way to use jit with optional arguments? This can be helpful if one has a number of jitted functions (e.g. different type of nonlinearities in a neural network) which accept the same fixed arguments but whose number of optional arguments differs. 

* is there a way for a jitted function to change the value of an optional argument in the function body? E.g., you pass an optional argument (of a fixed shape and data type) with a certain value and when the jitted function call is complete, the value of that variable passed into the argument has been updated?

* if a function calls in its body other sub-functions, is there any advantage to jit the separate sub-functions or does jitting the outermost function offer the same speed/advantages?

**Comment by clemisch** [comment]
> could someone clarify what causes this behavior?

The output shape of `mean` and ultimately your function depend on the value of the `axis` argument. For `jit` to work, the shapes of the output (and everything in between) may only depend on the shapes of the input arguments -- not on their values.

**Comment by clemisch** [comment]
> is there a way to use jit with optional arguments?

Yes, this is generally possible. Afaik it's not possible only for a 'static' argument (as is the case in your function) because functions with `static_argnums` have to be called with all positional arguments (see the docstring). I.e., the 'static' argument has to be positional, so can't be optional. That's the way I understand it -- I'd be interested in input from jax devs.

> is there a way for a jitted function to change the value of an optional argument in the function body?

I'm not sure if I understand you correctly, but jit'ed functions (or functions on `DeviceArrays` in general) don't have side effects, in that you can not change values of a `DeviceArray` in-place inside a function. So if you pass a `DeviceArray` to a function, the function will not alter this array. Maybe you could share a snippet to clarify what you mean.

> if a function calls in its body other sub-functions, is there any advantage to jit the separate sub-functions or does jitting the outermost function offer the same speed/advantages?

**Afaik**, no. If your code allows it, jit the outermost function. Jit'ing subfunctions may be viable if your outermost function has control flow that can not be jit'ed.

**Comment by mgbukov** [comment]
Thanks! 

About Q2 above: you did understand my question; I am aware than DeviceArrays cannot be updated in-place, which is how you would solve this in pure python. I was thinking along the lines of a cpp code, where I can define a variable outside my function, and then pass a pointer to that variable into that function to change the value at the memory address of the pointer inside the function body. Even though jax operates on a GPU, I don't see why this cannot be done. E.g., my variable can be stored on the CPU to avoid memory issues, etc; or if its size is not too big, it can also be stored on the GPU but then we need a way to access the values externally without the function returning that variable.

To give a concrete example: consider a `batchnorm` layer function which normalizes a minibatch of data to zero mean and unit variance. A batchnorm layer requires to actively compute the minibatch mean and variance during training, but it also has to passively use some fixed mean and variance during inference (so that we can evaluate in a meaningful way single-data point samples once training is complete), [https://arxiv.org/pdf/1502.03167.pdf](https://arxiv.org/pdf/1502.03167.pdf). 

I am trying to write a `batchnorm` function in jax which has a parameter to toggle the mode from `training` to `inference`. But the inference mode requires me to pass the values of the mean and variance computed during training. In python this is trivial: I pass the mean and variance and update their memory views in-place during `training` while using their fixed values directly for evaluation during `inference`. 

I am currently exploring ideas trying trick `jit` by passing a dictionary which contains the `mean` and `variance` to be updated during `training` but I'm not too optimistic it'll work. What I know works for sure is to do `training` without `jit` and use `jit` only for inference but this is not optimal.  

Another related Q: can `jit` be applied on entire classes? I've seen `numba` offers such an option but there the use of `jit` is fundamentally different as far as I understand.

**Comment by shoyer** [comment]
> > is there a way to use jit with optional arguments?
> 
> Yes, this is generally possible. Afaik it's not possible only for a 'static' argument (as is the case in your function) because functions with `static_argnums` have to be called with all positional arguments (see the docstring). I.e., the 'static' argument has to be positional, so can't be optional. That's the way I understand it -- I'd be interested in input from jax devs.

The standard way to handle this is to `jit` a private helper function -- see https://github.com/google/jax/issues/1925#issuecomment-569800183 for an example.

**Comment by shoyer** [comment]
> Another related Q: can `jit` be applied on entire classes? I've seen `numba` offers such an option but there the use of `jit` is fundamentally different as far as I understand.

No, you can't `jit` entire classes, but you can `jit` methods on classes (you will need to mark `self` with `static_argnums`). Keep in mind that if your classes have any mutable state that this will break, however.

**Comment by jekbradbury** [comment]
> To give a concrete example: consider a `batchnorm` layer function which normalizes a minibatch of data to zero mean and unit variance. A batchnorm layer requires to actively compute the minibatch mean and variance during training, but it also has to passively use some fixed mean and variance during inference (so that we can evaluate in a meaningful way single-data point samples once training is complete), https://arxiv.org/pdf/1502.03167.pdf.
> 
> I am trying to write a `batchnorm` function in jax which has a parameter to toggle the mode from `training` to `inference`. But the inference mode requires me to pass the values of the mean and variance computed during training. In python this is trivial: I pass the mean and variance and update their memory views in-place during `training` while using their fixed values directly for evaluation during `inference`.
> 
> I am currently exploring ideas trying trick `jit` by passing a dictionary which contains the `mean` and `variance` to be updated during `training` but I'm not too optimistic it'll work. What I know works for sure is to do `training` without `jit` and use `jit` only for inference but this is not optimal.

This is definitely something that's nontrivial to do in JAX, because `jax.jit` doesn't respect Python references/object identity (a kind of impurity or side effect in Python code). In general there are a few options: you can pass things like the running mean and variance as arguments (make the dataflow completely explicit), you can use impure sugar like in github.com/JuliusKunze/jaxnet to create the explicit-dataflow function, or you can use something in between like #1803 (where the dataflow is explicit but driven by stringy tags).

**Comment by mgbukov** [comment]
@jekbradbury I like the tagging feature from #1660 / #1803 and I think this should let me do exactly what I want, as I have an intermediate calculation  (the mean and variance), the values of which I want to access independently. But it looks like these changes are not yet merged into master, and I'd like to keep my code master-branch compatible. Any ideas on the timeframe you're planning to merge #1660 / #1803 into master?

**Comment by mattjj** [comment]
Predictions are hard, especially about the future, but we're working on it! I'll say "sometime this quarter" is likely.

If you add yourself to #1803 (e.g. by leaving a comment) you'll get notified as we make progress. 

In the meantime, can we close this issue? @mgbukov were your questions answered?

**Comment by mgbukov** [comment]
Yes, let me close this issue.
